pallets/loans-ref/src/lib.rs

```diff
@@ -0,0 +1,792 @@
+		/// The `amount` will be transferred from pool reserve to borrower.
+		/// The portfolio valuation of the pool is updated to reflect the new present value of the loan.
+		/// Rate accumulation will start after the first borrow.
+		#[pallet::weight(10_000)]
+		#[transactional]
+		pub fn borrow(
+			origin: OriginFor<T>,
+			pool_id: PoolIdOf<T>,
+			loan_id: T::LoanId,
+			amount: T::Balance,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			match CreatedLoans::<T>::take(pool_id, loan_id) {
+				Some(created_loan) => {
+					Self::ensure_loan_borrower(&who, &created_loan.borrower)?;
+					Self::make_active_loan(pool_id, loan_id, created_loan, |loan| {
+						loan.borrow(amount)
+					})?
+				}
+				None => Self::update_active_loan(pool_id, loan_id, |loan| {
+					Self::ensure_loan_borrower(&who, loan.borrower())?;
+					loan.borrow(amount)
+				})?,
+			};
+
+			T::Pool::withdraw(pool_id, who, amount)?;
+
+			Self::deposit_event(Event::<T>::Borrowed {
+				pool_id,
+				loan_id,
+				amount,
+			});
+
+			Ok(())
+		}
+
+		/// Transfers amount borrowed to the pool reserve.
+		///
+		/// The origin must be the borrower of the loan.
+		/// If the repaying amount is more than current debt, only current debt is transferred.
+		/// The borrow action should fulfill the borrow restrictions configured at [`types::LoanRestrictions`].
+		/// The `amount` will be transferred from borrower to pool reserve.
+		/// The portfolio valuation of the pool is updated to reflect the new present value of the loan.
+		#[pallet::weight(10_000)]
+		#[transactional]
+		pub fn repay(
+			origin: OriginFor<T>,
+			pool_id: PoolIdOf<T>,
+			loan_id: T::LoanId,
+			amount: T::Balance,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			let amount = Self::update_active_loan(pool_id, loan_id, |loan| {
+				Self::ensure_loan_borrower(&who, loan.borrower())?;
+				loan.repay(amount)
+			})?;
+
+			T::Pool::deposit(pool_id, who, amount)?;
+
+			Self::deposit_event(Event::<T>::Repaid {
+				pool_id,
+				loan_id,
+				amount,
+			});
+
+			Ok(())
+		}
+
+		/// Writes off an overdue loan.
+		///
+		/// This action will write off based on the write off policy configured by
+		/// [`Pallet::update_write_off_policy()`].
+		/// The write off action will only take effect if it writes down than the current write
+		/// off status of the loan. This action will never writes up.
+		/// No special permisions are required to this call.
+		/// The portfolio valuation of the pool is updated to reflect the new present value of the loan.
+		#[pallet::weight(10_000)]
+		#[transactional]
+		pub fn write_off(
+			origin: OriginFor<T>,
+			pool_id: PoolIdOf<T>,
+			loan_id: T::LoanId,
+		) -> DispatchResult {
+			ensure_signed(origin)?;
+
+			let status = Self::update_active_loan(pool_id, loan_id, |loan| {
+				let state = Self::find_write_off_state(pool_id, loan.maturity_date())?;
+				let limit = state.status().max(loan.write_off_status());
```

Comment:
> Why has the loan.write_off logic even this interface? Is it for admin?

Yes. The first parameter is the limit coming from the policy, next parameter is the value you want to apply. In this case, a simple write-off, will apply the same as the limit. If it was previously written off more than the policy because of an admin call, then it will try to write off the maximum (or penalty or percentage). i.e.

- write off admin: percentage: 0.5, penalty 0.2
- time passes
- write off with a policy that says: percentage 0.3, penalty 0.4
- The loan should be written off with the maximum: percentage 0.5, penalty 0.4

Reply:
This makes sense, but would be good to add a comment explaining this (maybe even just copy paste what you wrote here).